Introduce Pioneer into monorepo

Lezek123 opened the pull request:
This PR is an attempt to move Pioneer from the repository at https://github.com/Joystream/apps into this monorepo.

**Moving Pioneer into subfolder and merging**

I created a `monorepo-migration` branch in the Pioneer repository (https://github.com/Joystream/apps/tree/monorepo-migration), where I moved all the files into `pioneer` folder. This branch was then merged into `pioneer-migraton` (this one) using `--allow-unrelated-histories`

The entire flow was like this:
```
cd LOCAL_APPS_ROOT
git checkout joystream
git branch monorepo-migration
git checkout monorepo-migration
mkdir pioneer
ls -a1 | grep -v ^pioneer | xargs -I{} git mv {} pioneer
git commit -m "Move to subfolder"
git push upstream monorepo-migration

cd LOCAL_MONOREPO_ROOT
git checkout master
git remote add pioneer-upstream https://github.com/Joystream/apps.git
git fetch pioneer-upstream
git branch pioneer-migration
git checkout pioneer-migration
git merge pioneer-upstream/pioneer-migration --allow-unrelated-histories
```

**_The 3k+ commits in this PR are the result of moving the entire Pioneer history. All the configuration work described below is introducted in the last commit of this PR: https://github.com/Joystream/joystream/commit/643c08c7a04d096e66a33fb92b406ca90654643d_**

**Workspaces**

In order to have all current Pioneer's workspaces available among other monorepo workspaces, I landed on this configuration in the root directory `packages.json`:
```
	"workspaces": [
		"tests/network-tests",
		"cli",
		"types",
		"pioneer",
		"pioneer/packages/*"
	],
```
As you can see, I made a few adjustments there, ie.:

- Moved `joy-types` package from the Pioneer's `packages` directory into the root monorepo directory `types`. I think it doesn't make much sense to have `joy-types` nested inside Pioneer anymore. It is a quite universal library, which is used by many projects, like the CLI, query node, Atlas (?) etc. therefore I think it makes sense for it to be at least at the same level in the directory structure as all those other projects. It allso allows as to consume it inside Pioneer the same way it is consumed by other projects (ie. we have to specify `@joystream/types/lib` as an import, instead of `@joystream/types`.

- I removed the nested workspaces in `pioneer/packages.json` as I figured we don't need to define them twice if they are already defined in root `package.json` via `pioneer/packages/*`. It seems more clear and less error-prone to have all workspaces defined only once (in the root folder).

- Defined `cli` as one of the workspaces in order to test the effect of using shared `@joystream/types` among Pioneer and CLI.

The idea behid this structure is that we can run `yarn install` in the root directory in order to install and link dependencies for all the workspaces at once and then execute any command in Pioneer workspace ie. by running:

`yarn workspace pioneer COMMAND`
_Where COMMAND can be ie.: `build`, `start`, `storybook` etc._

Because Pioneer was very dependent on beeing a monorepo itself (instead of beeing part of a larger monorepo), this turned out not to be possible before making a few adjustments described below.

**Yarn, webpack and other issues**

In order to make Pioneer work as part of a larger monorepo (with commands like `start`, `build`, `storybook` and `lint` remaining usable) I needed to:

1 Fix the issue with `yarn.lock` beeing ignored

If Pioneer's lockfile is removed/ignored, yarn installs some conflicting versions of packages like `multicodec` (`v0.5` and `v1.0`), which causes Pioneer's `start` script to fail. When Pioneer becomes just a workspace in a monerepo, it's `yarn.lock` is no longer respected unless moved into the root directory. This forced me to include `yarn.lock` in the root folder as part of the commit.

It's probably not good that `yarn.lock` is required in order for the Pioneer to work, so we might want to investigate this issue further, this isn't strictly related to the process of moving into the monorepo though (it happens if we remove `yarn.lock` in the original repository too) and there were a lot of other issues that needed fixing as well, so I decided not to tackle this one now.

2. Pioneer doesn't build with newer TypeScript (ie. `3.8.3`)

If we don't specify the exact version of `typescript` in the Pioneer's `package.json`, we might get some newer version due to hoisting, which causes Pioneer build to fail. I fixed it by specifying `"typescript": "3.7.2"` in Pioneer's `package.json` (this is the version it was using when installed from it's own repository).

3. Some sub-dependency binaries not beeing linked in the workspace's `node_modules`

Binaries from some sub-dependency packages like `webpack`, `@babel/cli` etc. are not beeing linked in the `pioneer/node_modules/.bin` after running `yarn install` in the root directory unless those packages are specified as direct dependencies of Pioneer. This required adding packages like: `@babel/cli`, `@polkadot/dev`, `webpack`. `cpx` to Pioneer's `packages.json` (I tried using the same versions that would normally get installed).

4. Eslint plugins need to be required in Pioneer's `package.json` (cannot be found when hoisted)

There was some issue with `eslint` not beeing able to find plugins (which were sub-dependencies of ie. `@polkadot/dev-react`) hoisted to root `node_modules` also, so those needed to be specified as direct dev dependencies in Pioneer's `package.json` too. Those include:  `eslint-config-semistandard`, `eslint-config-standard`, `eslint-plugin-import`, `eslint-plugin-node`, `eslint-plugin-promise` and `eslint-plugin-standard`

5. Pioneer's `tsconfig` looking for files in `./node_modules`

Another issue was that due to relative path to some `node_modules` in Pioneer's `tsconfig.json`, some files could not be found by the TypeScript compiler. Those paths needed to be changed from `./node_modules` to `../node_modules`. 

6. `@joystream/types` vs `@joystream/types/lib`

Because `@joystream/types` was moved to separate workspace, independent from Pioneer, the imports from `@joystream/types` needed to be replaced to imports from `@joystream/types/lib`.  This was also mentioned in: https://github.com/Joystream/joystream/issues/321

I fixed it by executing following replace-by-regex's for all project files:

Replace `from '@joystream/types(?!/lib)` to `from '@joystream/types/lib`
Replace `from "@joystream/types(?!/lib)` to `from "@joystream/types/lib`
Replace `require\('@joystream/types(?!/lib)` to `require('@joystream/types/lib`

Strangely enough I noticed that, for example, the `cli` was succesfully using import from `@joystream/types` (without `lib`) before, even though it was always consuming `@joystream/types` externally. Perhaps it's due to `main` beeing specified in `types/package.json`. This may require some further investigating in order to make sure it's resolved correctly.  

7. The need for .yarnclean

I moved the `.yarnclean` file from `pioneer` folder into root folder.
I also needed to add `@polkadot/ts/node_modules` there, because for some reason Yarn was creating `node_modules` folder inside `@polkadot/ts`, which was causing TypeSciprt error and a failure while trying to build Pioneer.

Current `.yarnclean` looks as follows:
```
@types/react-native
@polkadot/ts/node_modules
```

8. `"types": ["node"]` in `@joystream/types` tsconfig

There was a need to add `"types": ["node"]` in `types/tsconfig.json`. Otherwise some type conflicts between `jest` and `mocha` were preventing `yarn workspace @joystream/types build` from running succesfully. 

9. `pioneer/.storybook/webpack.config.js` - needed to add ".js" extension

There was an error while trying to run `yarn workspace pioneer storybook` probably due to the fact that `@joystream/types` workspace was moved outside Pioneer and webpack wans't configured to process `.js` files that were coming from there.

**How to test**

While testing this branch I was so far mostly focused on those 3 things:
1. Do Pioneer commands run succesfully? (ie. `build`, `start`, `storybook`, `lint`)
2. Can I share `@joystream/types` workspace between Pioneer and CLI?
3. Does both `CLI` and `Pioneer` work correctly?

In order to initialize the monorepo (before trying to test anything) there are 2 steps required:

1. Run `yarn install` in the root directory (remove all `node_modules` before that if needed).
2. Run `yarn workspace @joystream/types build` (in order to build the types, so that the other workspaces can succesfully import them).

After that, commands like this should execute succesfully:
- `yarn workspace pioneer start`
- `yarn workspace pioneer build`
- `yarn workspace pioneer storybook`
- `yarn workspace pioneer lint`

In order to try CLI commands, you can run them like this (from the root directory):
`./cli/bin/run council:info`

Commit:
SetMaxValidatorCountForm

Commit:
TS props types fix

Commit:
Mint capacity form - split into Council and Content Working Group

Commit:
delete mocks

Commit:
fix merge conflict

Commit:
Add Types

Commit:
Runtime upgrade form

Commit:
Merge pull request #373 from Lezek123/set-storage-role-params

SetStorageRoleParamsForm + FormField

Commit:
Merge pull request #378 from Lezek123/content-working-group-mint

Proposal form: Content Working Group Mint Capacity (from MintCapacityForm)

Lezek123 committed:
Merge branch 'joystream-constantinople' into runtime-upgrade

Commit:
Merge pull request #379 from Lezek123/runtime-upgrade

Proposal form: Runtime upgrade

Commit:
Change return type of hasVoted

Commit:
Add Basic Index To Proposals

Commit:
Refactorig

Commit:
Add Proposal Tab in App

Commit:
Merge pull request #377 from Lezek123/set-max-validator-count

Proposal form: SetMaxValidatorCount

Commit:
Merge pull request #382 from Lezek123/proposal-forms

Proposal forms: Refactoring

mnaamani committed:
Merge branch 'joystream-constantinople' into add-missing-types

Commit:
Merge pull request #362 from mnaamani/add-missing-types

add missing types to @joystream/types

Commit:
Solved conflict

Commit:
Add Transport

Commit:
Choose proposal type view

Commit:
More mock data

Commit:
Add Methods to get ProposalTypes Details

Commit:
Types in most components have been adjusted. Some bugs corrected and
some variables got a bit stronger typing. The build errors went down to
18 from 40.

Commit:
Bugfix

Commit:
Merge with joystream-constantinoples

Commit:
Merge branch 'francesco-transport-runtime' into francesco-add-context

Commit:
Merge pull request #383 from Lezek123/choose-proposal-type

Choose Proposal Type screen

Commit:
Fix Votes.tsx bug

Commit:
Merge branch 'francesco-proposal-index' into francesco-add-context

Commit:
Rename proposalEngine to proposalsEngine

Commit:
Correct typo in transport

Commit:
Add Fetch Logic in ProposalPreviewList

Commit:
Add Loading, Error and ProposalFromId Components

Commit:
Refactor excludeKeys to includeKeys

Commit:
Fix Typo in index.tsx

Commit:
Fix inconsistencies with joystream-constantinoples

Commit:
Bring branch up to date with francesco-transport

Commit:
Bring branch up to date with remote

Commit:
Will amend

Commit:
Code Review Corrections

Commit:
Remove duplicate Route

Commit:
Merge pull request #380 from Gamaranto/francesco-proposal-index

Francesco proposal index

Commit:
Add Proposal Type

Commit:
Merge branch 'joystream-constantinople' of github.com:Joystream/apps into francesco-add-missing-types

Commit:
Add ProposalOf Type

Commit:
Bring branch up to date with remote

Commit:
Proposal forms: Extrinsics

Commit:
Add CalculateStake Util

Commit:
This commit changes the proposalTypesVotingPeriod and
proosalTypesGracePeriod to return an object instead of an array with the
results. Fetching before was dependent on order while now a key is the
only thing needed to fetch data.

Commit:
CleanUp

Commit:
Add Proposal Details Type

Commit:
Route names fix

Commit:
Fix Proposal Type

Commit:
Add Missing Types

Commit:
Move includeKeys to utils

Commit:
Add CleanProposalType

Commit:
Add ParsedProposal Type

Commit:
Refactor ProposalPreview List and ProposalFromId

Commit:
Merge pull request #386 from Lezek123/forms-extrinsics

Proposal forms: Extrinsics

Commit:
Bring up to date with base branch

Commit:
Add VoteByProposalAndMember and additional Types and Methods

Commit:
Add Votes Method

Commit:
Bugfix

Commit:
Improve formatting of votes method

Commit:
Add method to fetch a proposal Type Metadata.

Commit:
Better define ProposalType

Commit:
Fix Error in usePromise

Commit:
ProposalDetails, ProposalList etc. - transport integration

Commit:
Add Slugify Method

Commit:
Refactor some logic to "getExtendedStatus" & voting section display conditions

Commit:
ChooseProposalType now shows up

Commit:
This commit refactors some files to fix build errors and improves
readability. The most notable change is the splitOnUpperCase function in
the utils package, used in multiple places in the transport.

Commit:
Code Review Changes

Commit:
Merge branch 'francesco-add-context' into transport-integration

Commit:
Merge branch 'francesco-add-context' into transport-integration

Commit:
Remove DateFromBlock and Add Proper Block Date Derivation

Commit:
Fix routes

Commit:
Add Exact to Text Route

Commit:
Add Route Fix

Commit:
Merge branch 'francesco-add-context' into transport-integration

Commit:
Votes - basic integration

Commit:
Merge pull request #385 from Gamaranto/francesco-add-context

Francesco add context [WIP]

Commit:
Merge branch 'joystream-constantinople' into transport-integration

Commit:
Remove obsolete comment

Commit:
Fix build errors

Commit:
Minor formatting fixes

Commit:
Merge pull request #387 from Lezek123/transport-integration

Transport integration

Commit:
Add Validation Schema

Commit:
Real data for forms

Commit:
Build fix

Commit:
Vote kinds enum order fix

Commit:
Untangle Validation Logic From Forms

Commit:
Merge pull request #389 from Lezek123/vote-kinds-enum-fix

Vote kinds enum order fix

Commit:
Merge pull request #388 from Lezek123/forms-data

Fetch data for forms

Commit:
Bring up to date with base branch

Commit:
Improve Typing

Commit:
Proposal statuses fix

Commit:
Make Validation Values Dynamic

Commit:
Add Error Messages

Commit:
Add Validation to Mint Capacity Form.
While being a 1 line change, this commit "kills" the MintCapacityForm
abstraction by forcing it with to specific validation. While this is not
a breaking change by any means (we don't have others MintForms in this
release) it forces a (small) abstraction to be less useful. Comments and
feedback are welcomed even if I would leave this as is for the current
release.

Commit:
Align pic and handle in Details

Commit:
Align pic and handle in Votes

Commit:
Fix text-align in Proposal Params

Commit:
Re-add Icon to Submit Button

Commit:
Fix Backer Typo

Commit:
Fix Typo

Commit:
Change File Validation to Mixed

Commit:
Refactor imports and fix build errors

Commit:
Improved ProposalDetails data flow

Commit:
Merge pull request #391 from Lezek123/proposal-statuses

Proposal statuses fix

Commit:
Merge pull request #394 from Lezek123/proposal-details-data-flow

Improved proposal details data flow

Commit:
Proposal withdrawal

Commit:
ActiveStake.stakeId type fix

Commit:
Merge pull request #396 from Lezek123/active-stake-type-fix

ActiveStake.stakeId type bug fix

Commit:
Merge branch 'joystream-constantinople' into francesco-better-validation

Commit:
Proposal cancellation - message

Commit:
Add Real Values for Validation and Unit To Error Messages.

Commit:
Fix Build Error

Commit:
Fix units and some values

Commit:
Fix units and some values

Commit:
ReAdd Filtering Options

Commit:
Update Values and Bugfix

Commit:
Fix Build Errors

Commit:
Add Integer Check to All Fields

Commit:
Bugfix

Commit:
Merge pull request #393 from Gamaranto/francesco-council-tab

Fix Council Tab crash

Commit:
Merge pull request #390 from Gamaranto/francesco-better-validation

Francesco better validation

Commit:
Refactor filterProposals and Menu

Commit:
Merge pull request #395 from Lezek123/proposal-withdrawal

Proposal withdrawal implementation

Commit:
Rename Withdrawn to Canceled

Commit:
Merge pull request #398 from Gamaranto/francesco-filter-proposals

Francesco filter proposals

Commit:
Add cancellation fee fetched from runtime

Commit:
Minor build fix

Commit:
Extract ProfilePreview Component

Commit:
Merge branch 'joystream-constantinople' into francesco-fix-styles

Commit:
Election parameters defaults + hide irrelevant

Commit:
WG mint cap default

Commit:
Set lead default & info

Commit:
Council params form - placeholders, re-enabling limit and size fields

Commit:
Merge pull request #392 from Gamaranto/francesco-fix-styles

Francesco fix styles

Commit:
Merge pull request #399 from Lezek123/fetch-cancellation-fee

Cancellation fee: fetch from rutime

Commit:
SetLead proposal details profile preview

Lezek123 committed:
Merge branch 'joystream-constantinople' into forms-defaults

Commit:
Minor conflit resolve fix

Commit:
Merge pull request #400 from Lezek123/forms-defaults

SetElectionParameters, SetContentWorkingGroupLead, SetContentWorkingGroupMintCap - fetch current data from runtime

Commit:
Merge pull request #401 from Lezek123/set-lead-profile-preview

SetLead proposal details member profile preview

Commit:
Add MaxValidator Count Placeholders

Commit:
Add SetStorageRoleParams Placeholders and fix inconsistency in Validation

Commit:
Fix Build Errors

Commit:
Hide not enforced parameters

Commit:
Revert Min Actors Max to 1

Commit:
Update the stakes

Commit:
Merge pull request #402 from Gamaranto/francesco-hide-input-fields

Hide input fields and add placeholders for SetStorageRoleParameters and SetMaxValidator count.

Commit:
Merge pull request #404 from Lezek123/hide-not-enforced

Hide not enforced parameters on proposal details page

Commit:
Merge pull request #407 from Lezek123/update-stakes

Update proposals' stakes

Commit:
Fix block hash query

Commit:
Merge pull request #408 from Lezek123/fix-block-hash-query

Fix block hash query

Commit:
Memberships - pagination

Commit:
Add voting period & fix build errors

Commit:
Set lead - allow 'NONE'

Commit:
Optimalizations

Commit:
Word wrapping for rationale and parameters

Commit:
Grace period: Expires in => Executres is

Commit:
Merge pull request #421 from Lezek123/executes-in

Grace period: Expires in => Executres is

Commit:
Merge pull request #414 from Lezek123/voting-period-on-type-preview

Voting period on ProposalTypePreview

Commit:
Merge pull request #418 from Lezek123/description-wrapping

Word wrapping in proposal description and parameters

Commit:
Merge pull request #415 from Lezek123/set-empty-lead

Allow setting empty lead + members dropdown optimalizations

Commit:
Minor UX/UI improvements

Commit:
Merge 'joystream-constantinople' and fix conflicts

Commit:
Build fix + minor css fix

Commit:
Details.tsx - block heights and execution error

Commit:
Add "Expired" proposals tab

Commit:
Quorum and threshold on ProposalType preview

Commit:
Build errors etc

Commit:
Remove hardcoded "tJOY" references

Commit:
lastId formula fix

Commit:
Proposal id in overview

Commit:
Member loading bug fix

Commit:
Election params validation fix

Commit:
Merge pull request #426 from Lezek123/block-heights

Details.tsx - block heights and execution error

Commit:
Merge pull request #427 from Lezek123/expired-tab

Add "Expired" proposals tab

Commit:
Merge pull request #431 from Lezek123/tjoy-renaming

Remove hardcoded "tJOY" references

Commit:
Merge pull request #423 from Lezek123/ui-ux-glitches

Minor UX/UI improvements

Commit:
Merge pull request #436 from Lezek123/election-params-validation-fix

Election params validation fix

Commit:
Merge pull request #433 from Lezek123/proposal-id-overview

Proposal id in overview box

Lezek123 committed:
Merge branch 'joystream-constantinople' into quorum-thrershold

Commit:
Council - identify by membership (step 1)

Commit:
Member profile - optional id

Commit:
Minor loader fix

Commit:
Members dropdown

Commit:
Merge pull request #429 from Lezek123/quorum-thrershold

Quorum and threshold in ProposalTypePreview

deamme committed:
Remove unnecessary tabs

deamme committed:
Added ability to set custom endpoint via console

deamme committed:
Remove needless tabs and text

bedeho committed:
Merge pull request #442 from deamme/issue-439

Tab hell fix

mnaamani committed:
revert forum type ThreadId to u64

mnaamani committed:
joy-types: bump to version 0.8.0 for Constantinople release

mnaamani committed:
joy-types: remove temp file

Commit:
Fixes

mnaamani committed:
u64 ThreadId

Commit:
Merge 'joystream' and fix conflicts

Commit:
Final minor touch

Commit:
Merge pull request #411 from Lezek123/memberships-pagination

Memberships - pagination

deamme committed:
Stop hiding information from people without a key

mnaamani committed:
joy-types: add missing type ThreadCounter from proposal discussion module

mnaamani committed:
Merge branch 'joystream-constantinople' into joy-types-bump-version

mnaamani committed:
joy-types: v0.9.0

mnaamani committed:
Merge pull request #443 from mnaamani/fix/thread-id

revert forum type ThreadId to u64

mnaamani committed:
Merge remote-tracking branch 'joystream/joystream-constantinople' into joy-types-bump-version

mnaamani committed:
Merge pull request #445 from Lezek123/joy-media-fixes-and-status-messages

Joy media fixes and status messages

mnaamani committed:
Merge branch 'joystream' into alex-rename-link-to-links-updated

deamme committed:
Fixed issues based on review

deamme committed:
Fixed issues based on second review

deamme committed:
Fixed issues based on third review

Lezek123 committed:
Merge pull request #446 from deamme/issue-440

Stop hiding information from people without a key

mnaamani committed:
joy-types: add missing lodash dependency and bump version to v0.9.1

Commit:
Merge pull request #444 from mnaamani/joy-types-bump-version

Bump @joystream/types package

Commit:
Merge 'joystream' and fix the conflicts

deamme committed:
Add transfer tab back

Commit:
Rename to withOnlyAccounts for consitency

Lezek123 committed:
Merge pull request #454 from deamme/add-transfer-back

Add transfer tab back

Commit:
Change links and remove faucet references

Commit:
Merge 'joystream-constantinople' and fix conflicts

Commit:
Rename default WS endpoint

Commit:
Change "Tokenomics" link

Lezek123 committed:
Merge pull request #455 from Lezek123/joystream-conflicts-fix

Joystream conflicts fix

Commit:
Reintroduce extrinsics, settings, storage and sudo

mnaamani committed:
Merge pull request #458 from Lezek123/reintroduce-api-tabs

Reintroduce extrinsics, settings, storage and sudo

Lezek123 committed:
Merge branch 'joystream-constantinople' into references-to-older-chains

Commit:
Reintorduce explorer + back to light/full mode

Commit:
JavaScript, Toolbox

Lezek123 committed:
Merge pull request #457 from Lezek123/references-to-older-chains

References to older chains

Commit:
Remove outer links, ToS and PP from side menu

Commit:
Membership hooks - add isSubscribed

shamil-gadelshin committed:
Merge pull request #445 from Joystream/development

Constantinople Release

mnaamani committed:
Merge pull request #459 from Lezek123/reintroduce-tabs-continuation

Reintorduce explorer and bring back to "light"/"full" mode

Lezek123 committed:
Merge pull request #438 from Lezek123/council-members-identity

Council - identify candidates/members by membership

bwhm committed:
Merge pull request #453 from Joystream/joystream-constantinople

Constantinople Deployment

Commit:
Encoding fix + tmp rename "Links" to "Link" in UI

Commit:
Force description too

Commit:
Moving changes back to schema

mnaamani committed:
Merge pull request #461 from Lezek123/media-schemas-and-links-fix

Media schemas and links fix

Commit:
Move to subfolder

Commit:
Merge remote-tracking branch 'pioneer-upstream/monorepo-migration' into pioneer-migration

Commit:
Pioneer workspaces config etc.

mnaamani commented:
Brilliant work on this migration.

I hope when/if we pull from upstream `polkadot-js/apps` merge conflicts will not be too difficult to deal with, and I'm sure documenting the changes after import here will come in handy.

The only issue I have is that `@joystream/types` still transpiles into the `lib/` folder and the import path used by consuming apps still have to contain this `/lib`. I always felt like I made a bad choice with that approach and was hoping to make it more uniform with all the other packages in pioneer where code is transpiled into `build/` and somehow when importing the "build/" is not required in the path. Is this some webpack magic or different typescript/babel configuration done for other packages?

Can we improve on `joystream/types` in a similar way?

When working in old Pioneer repo, one didn't need to think about having to build dependencies before the main build, it was automatic.

Here as you pointed out we have to remember to rebuild @joystream/types and then the project that depends on it. I'm guess webpack was always taking care of rebuilding dependencies so you could just modify types and run yarn start and code would be automatically transpiled.

Is there a way to get a similar behavior in this new repo. What I'm imagining is that after someone clones the repo, they just need to run `yarn` and it will install and build all dependencies including any local ones and build all our apps, so the next step is just to run one of the apps.

I see you are using  `"*"` as the version for the `@joystream/types` package in the `cli` pacakge.json will this not be a problem when/if we choose to publish the cli for example to npm registry. 

We should remove pioneer/yarn.lock and cli/yarn.lock now that only the top .lock file is required.

I don't quite understand the resolutions section in the package.json but doesn't the one in pioneer/package.json need to be moved to the root package.json? It does seem to be relevant to any app that is depending on the polkadot/api to ensure they all use the same versions.

mnaamani commented:
I was able to build pioneer and also run it with development `yarn workspace pioneer start` works well.

Just wondering why is `@joystream/types` added in `devDependencies` and not just `dependencies` ?

Commit:
Minor fixes / cleanup

Commit:
Pioneer webpack/tsconfig for "@joystream/types"

Lezek123 commented:
I wasn't able to make it all work the way we would want for now, but at least manged to make the workflow with Pioneer as simple as it was before, while maintaining `types` in the root folder and easily accessible for other projects.

I configured `pioneer/packages/apps/webpack.config.js` and `pioneer/tsconfig.js` so that now they are aware of the new location of `@joystream/types` and all the changes are processed on-fly when running `yarn workspace pioneer start` (same way it was functioning before the merge into monorepo), so we don't need to "manually" build `@joystream/types` anymore.

Optimally we would want something like that for the CLI (and potentially other projects) too, but I couldn't find a way to make it work for now, so the CLI and `tests/network-tests` still rely on the build version of `@joystream/types`.

Having that in mind I added `yarn workspace @joystream/types build` as a `postinstall` script in the root `package.json`, so at least we can work on those tools right after running `yarn install`. We just need to remember to re-build the `@joystream/types` if we decide to change them along the way.

One more thing to keep in mind is that this gets us back to the disparity of how `@joystream/types` are "consumed" by Pioneer vs other projects. The other projects still need to add the `/lib` if importing from other files/folders than `@joystream/types/lib/index.js`.

I think this may be a good-enough setup for now, it could definitely be improved, but that would require some more research on my part and we'd have to probably change the way we're handling the process of building and publishing `@joystream/types` etc.

I also resolved a few of the smaller issues mentioned:
- `@joystream/types` beeing a `devDependency` by mistake
- Using `"*"` for the version of `@joystream/types`
- Moved the "resolutions" section from Pioneer to root `package.json`

mnaamani committed:
Merge branch 'development' into pioneer-migration
